### PR TITLE
Add meson + ninja support

### PR DESCRIPTION
**Pull request by th0ma7**
_Motivation:_  Some newer package version mandatory require meson + ninja native build environment.
_Linked issues:_  N/A

### Checklist
- [ ] Build rule `all-supported` completed successfully
- [ ] Package upgrade completed successfully
- [ ] New installation of package completed successfully

### GLIB Specific
- Known build issue starting with 2.63.4
- Last known version to work on `hi3535` (`armv7l`): 2.63.3 (unstable)
- Issue created at:  https://gitlab.gnome.org/GNOME/glib/-/issues/2274

### Migrated `cross/package`
- [x] `libdri`
- [x] `libsigc++`
- [x] `glib`
- [x] `glibmm`
- [x] `pixman`
- [x] `zstd`

### CODE REVIEW

File: mk/meson/ppc64.cfg [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+c_args = ['']
+c_link_args = ['']
+cpp_args = ['']
+cpp_link_args = ['']
+
+[host_machine]
+system = 'linux'
+cpu_family = 'ppc64'
+cpu = 'ppc64'
+endian = 'little'
```
Comment (hgy59) [inline mk/meson/ppc64.cfg:11]: the active ppc archs (ppc853x and qoriq) are both 32 bit, so I expect ppc32 here.
and the endian is 'big' not 'little'.

**Comment by hgy59** [comment]
@th0ma7 good work so far! 
I hated meson when I first tried to cross compile packages here. 
See my comment for ppc archs. I have a DS210+ and could test ppc853x-5.2 packages.

File: mk/meson/ppc64.cfg (hunk 1 above)
Comment (th0ma7) [reply mk/meson/ppc64.cfg]: now changed.

**Comment by th0ma7** [comment]
> @th0ma7 good work so far!
> I hated meson when I first tried to cross compile packages here.
> See my comment for ppc archs. I have a DS210+ and could test ppc853x-5.2 packages.

@hgy59 thnx for looking into this.  Indeed, testing on each platforms is mandatory so any help I can get on this is appreciated. 

It took me quite a few cycles in order to get into "something" working.  And I'd really much like to get rid of the configuration files and better understand the various options (and how to translate the cross-files into options).

Note that I was able to compile the latest glib yesterday with meson+ninja with almost not efforts and close to empty `makefile`... So with luck this might end-up working out all right :)

**Comment by ymartin59** [comment]
Do I understand well you had to natively compile dedicated cmake version? What are troubles with cmake from Debian docker image?
I am ready to prepare a new Debian 10 / Buster image if it prevents such addition in framework.

**Comment by th0ma7** [comment]
@ymartin59 actually, its Debian 10 `cmake` version that is too old so I added a native build to circumvent this.
To circumvent this I could probably use an older version of `ninja`...

I've updated my own `lxc` image to Debian 10 a while ago, perhaps we should make a switch to Ubuntu 20.04?
I can share the doc as needed, also have it for Debian 11/Ubuntu 20.04 as only a handful of package changes between versions.
The only package that refused to build was `gnuTLS` but I updated it since so it might be all good now.

This is a `[WIP]` so I expect things to evolve before it's in a readiness state.

**Comment by th0ma7** [comment]
Also, current build error on x86_64 is based on the lack of meson being installed on the build image.
Current build image requires a `$ lxc exec $image -- /bin/bash -c "pip3 install meson"`

**Comment by hgy59** [comment]
> 
> 
> Also, current build error on x86_64 is based on the lack of meson being installed on the build image.
> Current build image requires a `$ lxc exec $image -- /bin/bash -c "pip3 install meson"`

I propose to create a separate PR for the Dockerfile update. After mergin that, the new image will be published to docker hub and will be used for the github build action.

**Comment by ymartin59** [comment]
Debian buster-backports provide cmake 3.16.3 - is that version recent enough for ninja? if so I propose to add backports source to build image Dockerfile

**Comment by th0ma7** [comment]
> Debian buster-backports provide cmake 3.16.3 - is that version recent enough for ninja? if so I propose to add backports source to build image Dockerfile

Give me a few cycles and I'll play a bit with it and see.  I might also just use `ninja` with dependencies from backports as well which would lead into many changes in my current code.

**Comment by th0ma7** [comment]
@ymartin59 I redid some tests and the default `ninja` ends-up working out just well.  My first impressions where wrong and caused by the relatively old version of `meson` that Debian 10 provides.  But using `pip3` installation of meson provides a much newer version and solves all issues associated with default `ninja`.

All to say that building `native/cmake` and `native/ninja` is not necessary anymore and was removed entirely.  Although the benefit ended-up being the creation of spksrc.native-cmake*.mk for future usages.

@hgy59 I will be sending a parallel PR for adding up `ninja-build` package to the default Debian 10 image.  Note that this will cause the image to have 2x version of meson installed due to dependencies:
* `/usr/bin/meson` - default installed via apt at version 0.49
* `/usr/local/bin/meson` - locally `pip3` installed version being 0.56

Default will always fall-back to `/usr/local/bin/meson`.

**Comment by ymartin59** [comment]
In that case, why not remove Debian "meson" package from Docker image to avoid future troubles?

**Comment by ymartin59** [comment]
@th0ma7 Why not keep "native/cmake" and mk/spksrc.native-cmake*.mk files in a dedicated "standby" branch in spksrc - in case it may become useful one day? (then I would not include mk/spksrc.native-cmake*.mk in that PR)

**Comment by th0ma7** [comment]
> In that case, why not remove Debian "meson" package from Docker image to avoid future troubles?

I thought `ninja-build` required `meson` as depedendcy, but seems not... So I guess meson indeed will not be triggered for installation when invoking the install for `ninja-build` :
```
$ apt-cache showpkg ninja-build
Package: ninja-build
Versions: 
1.8.2-1 (/var/lib/apt/lists/deb.debian.org_debian_dists_buster_main_binary-amd64_Packages) (/var/lib/dpkg/status)
 Description Language: 
                 File: /var/lib/apt/lists/deb.debian.org_debian_dists_buster_main_binary-amd64_Packages
                  MD5: 5ef1371c63de86fcf36c1ddc875d2f26
 Description Language: 
                 File: /var/lib/apt/lists/deb.debian.org_debian_dists_buster_main_binary-i386_Packages
                  MD5: 5ef1371c63de86fcf36c1ddc875d2f26
 Description Language: en
                 File: /var/lib/apt/lists/deb.debian.org_debian_dists_buster_main_i18n_Translation-en
                  MD5: 5ef1371c63de86fcf36c1ddc875d2f26


Reverse Depends: 
  meson,ninja-build 1.6
  ninja-build:i386,ninja-build
  polymake,ninja-build
  cmake,ninja-build
  kdevelop,ninja-build
Dependencies: 
1.8.2-1 - libc6 (2 2.15) libstdc++6 (2 5.2) ninja (0 (null)) ninja-build:i386 (32 (null)) 
Provides: 
1.8.2-1 - ninja-build:i386 (= 1.8.2-1) 
Reverse Provides: 
ninja-build:i386 1.8.2-1 (= 1.8.2-1)
```

**Comment by th0ma7** [comment]
> @th0ma7 Why not keep "native/cmake" and mk/spksrc.native-cmake*.mk files in a dedicated "standby" branch in spksrc - in case it may become useful one day? (then I would not include mk/spksrc.native-cmake*.mk in that PR)

I'm ok at separating them into another PR.  Although note that `native/ninja` required `mk/spksrc.native-cmake*.mk` in conjunction with a more recent `cmake`.  The `native/cmake` only required regular autotools.  I would suggest only keeping the `mk` files while removing this section of code referring to `native/cmake` which as no use anymore :
```
# Use native cmake
DEPENDS += native/cmake
CMAKE_PATH = $(WORK_DIR)/../../../native/cmake/work-native/install/usr/local/bin
ENV += PATH=$(CMAKE_PATH):$$PATH
```

**Comment by th0ma7** [comment]
@hgy59 I think I nailed the last remaining blocker on this PR (besides ppc853x now being incompatible with the newer `glib`).  Although your massive PR #4320 will have a direct impact over this one (or vice-versa).
- Will there be a need for arch specific profile for ARMv7L for hi3535 arch?  Currently I only have a mk/meson/armv7.cfg file.
- Also I'll need help from people to confirm build outputs do work on various arches... so I can make de necessary adjustments on the arch profiles as needed (if any)

**Comment by hgy59** [comment]
>     * Will there be a need for arch specific profile for ARMv7L for hi3535 arch?  Currently I only have a mk/meson/armv7.cfg file.

ARMv7L has no hardfloat (floatingpoint instructions) included. Here you find some information about the `-mfloat-abi` (not only for ARMv5)
https://github.com/mesonbuild/meson/issues/5973

**Comment by th0ma7** [comment]
There seems to be a `cpu = 'armv7l'` option.  Made a few changes to simplify things and manage hi3535 as armv7l already (to be migrated to ARMv7L_ARCHS eventually).  Let's see how the build goes.

**Comment by th0ma7** [comment]
@hgy59 and @ymartin59 all packages using meson have now been migrated towards it.  Everything builds OK and much faster in many cases.  Build failures in `ppc853x` are related to unsupported `cross/package`.

I believe this PR is ready for merge but would clearly benefit from a last round of review.

Cheers and Merry Christmas!

File: cross/zstd/Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 PKG_NAME = zstd
 # warning: 1.4.4 and 1.4.5 fail for arch-aarch64
```
Comment (hgy59) [inline cross/zstd/Makefile:2]: comment seems obsolete now

File: mk/meson/armv8.cfg [hunk 3]
```diff
@@ -0,0 +1,8 @@
+[built-in]
+cpp_args = ['-fPIC']
+
+[host_machine]
+system = 'linux'
+cpu_family = 'aarch64'
+cpu = 'aarch64'
```
Comment (hgy59) [inline mk/meson/armv8.cfg:7]: for consistency this file could be renamed to aarch64.cfg (file name = cpu)

File: mk/spksrc.native-meson-env.mk [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# meson cross-compilation definitions
+MESON_CFG=$(WORK_DIR)/../../../mk/meson
```
Comment (hgy59) [inline mk/spksrc.native-meson-env.mk:2]: might be named `MESON_CFG_DIR`

File: native/glib/Makefile [hunk 5]
```diff
@@ -1,26 +1,25 @@
 PKG_NAME = glib
-PKG_SHORT_VERS = 2.58
-PKG_VERS = $(PKG_SHORT_VERS).3
+PKG_SHORT_VERS = 2.62
+PKG_VERS = $(PKG_SHORT_VERS).6
```
Comment (hgy59) [inline native/glib/Makefile:3]:
```suggestion
PKG_VERS = 2.62.6
PKG_SHORT_VERS = $(word 1,$(subst ., ,$(PKG_VERS))).$(word 2,$(subst ., ,$(PKG_VERS)))
```

File: native/glib/Makefile [hunk 6]
```diff
@@ -1,26 +1,25 @@
...
 PKG_EXT = tar.xz
 PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
 PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
 PKG_DIST_SITE = https://ftp.gnome.org/pub/GNOME/sources/$(PKG_NAME)/$(PKG_SHORT_VERS)
 
-DEPENDS =
+DEPENDS = 
+#DEPENDS = cross/libffi cross/pcre cross/zlib
```
Comment (hgy59) [inline native/glib/Makefile:10]: obsolete comment?

File: native/glib/Makefile [hunk 7]
```diff
@@ -1,26 +1,25 @@
...
-GNU_CONFIGURE = 1
+UNSUPPORTED_ARCHS = powerpc ppc824x ppc853x ppc854x
 
-PRE_CONFIGURE_TARGET = glib_pre_configure
+# https://developer.gnome.org/glib/
+# https://developer.gnome.org/glib/stable/glib-building.html
+# 2.62.6, 2.63.3-OK, 2.63.4-ERROR
+CONFIGURE_ARGS += -Dman=false -Dlibmount=false -Ddefault_library=shared -Dselinux=disabled
+# 2.63.5, 2.66.3
+#CONFIGURE_ARGS += -Dman=false -Dlibmount=disabled -Ddefault_library=shared -Dselinux=disabled
```
Comment (hgy59) [inline native/glib/Makefile:23]: same as for cross/glib

~but, can we remove the whole package native/glib?~

**Review by hgy59 — COMMENTED** [review]
Wow, a huge work!
Just some minor suggestions in my view...

File: mk/meson/x86.cfg [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[built-in]
+c_args = ['-m32']
+c_link_args = ['-m32']
+cpp_args = ['-m32']
+cpp_link_args = ['-m32']
+
+[host_machine]
+system = 'linux'
+cpu_family = 'x86'
+cpu = 'i686'
```
[the remark this answers is not in the document]
Comment (th0ma7) [reply mk/meson/x86.cfg]: done

File: mk/spksrc.cross-meson-env.mk [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  CONFIGURE_ARGS += --cross-file $(MESON_CFG)/armv7l.cfg
+endif
+ifeq ($(findstring $(ARCH),$(ARM8_ARCHES)),$(ARCH))
+  CONFIGURE_ARGS += --cross-file $(MESON_CFG)/armv8.cfg
+endif
+ifeq ($(findstring $(ARCH), $(PPC_ARCHES)),$(ARCH))
+  CONFIGURE_ARGS += --cross-file $(MESON_CFG)/ppc.cfg
+endif
+ifeq ($(findstring $(ARCH),$(x86_ARCHES)),$(ARCH))
+  CONFIGURE_ARGS += --cross-file $(MESON_CFG)/x86.cfg
```
[the remark this answers is not in the document]
Comment (th0ma7) [reply mk/spksrc.cross-meson-env.mk]: done

File: cross/glib/Makefile [hunk 10]
```diff
@@ -1,49 +1,24 @@
 PKG_NAME = glib
-PKG_SHORT_VERS = 2.58
-PKG_VERS = $(PKG_SHORT_VERS).3
+PKG_SHORT_VERS = 2.62
+PKG_VERS = $(PKG_SHORT_VERS).6
```
[the remark this answers is not in the document]
Comment (th0ma7) [reply cross/glib/Makefile]: I find this a overwhelming.  At this point we'd be better off with a versioning wrapper that would provide `PKG_MAJOR`, `PKG_MINOR` or other.  I guess for some other time.

For the moment, consider that done.

File: cross/glib/Makefile [hunk 11]
```diff
@@ -1,49 +1,24 @@
...
 PKG_EXT = tar.xz
 PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
 PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
 PKG_DIST_SITE = https://ftp.gnome.org/pub/GNOME/sources/$(PKG_NAME)/$(PKG_SHORT_VERS)
 
-DEPENDS = native/glib cross/libffi cross/pcre cross/zlib
```
[the remark this answers is not in the document]
Comment (th0ma7) [reply cross/glib/Makefile:9]: humm...Looking good to me, it show:
```
- DEPENDS = native/glib cross/libffi cross/pcre cross/zlib
+ DEPENDS = cross/libffi cross/pcre cross/zlib`
```

File: cross/glib/Makefile [hunk 12]
```diff
@@ -1,49 +1,24 @@
...
-CONFIGURE_ARGS += GLIB_COMPILE_RESOURCES=$(WORK_DIR)/../../../native/glib/work-native/$(PKG_DIR)/gio/glib-compile-resources
+UNSUPPORTED_ARCHS = powerpc ppc824x ppc853x ppc854x
 
-ADDITIONAL_CFLAGS = -Wno-unused-value
+# https://developer.gnome.org/glib/
+# https://developer.gnome.org/glib/stable/glib-building.html
+# 2.62.6, 2.63.3-OK, 2.63.4-ERROR
+CONFIGURE_ARGS += -Dman=false -Dlibmount=false -Ddefault_library=shared -Dselinux=disabled
+# 2.63.5, 2.66.3
+#CONFIGURE_ARGS += -Dman=false -Dlibmount=disabled -Ddefault_library=shared -Dselinux=disabled
```
[the remark this answers is not in the document]
Comment (th0ma7) [reply cross/glib/Makefile:22]: nah, I really want to keep it.  It happen that `glib` fails to build starting with version 2.63.4 and **only** on hi3535 arch.
I opened-up a bug request at https://gitlab.gnome.org/GNOME/glib/-/issues/2274
So commented `CONFIGURE_ARGS` is ready for next round of update.

File: mk/spksrc.native-meson-env.mk (hunk 4 above)
Comment (th0ma7) [reply mk/spksrc.native-meson-env.mk]: done

File: mk/meson/armv8.cfg (hunk 3 above)
Comment (th0ma7) [reply mk/meson/armv8.cfg]: good idea, done.

File: native/glib/Makefile (hunk 7 above)
Comment (th0ma7) [reply native/glib/Makefile]: indeed.  Kept the `cross/glib` but removing duplicate comments in `native/glib`.

File: native/glib/Makefile (hunk 6 above)
Comment (th0ma7) [reply native/glib/Makefile]: done

**Comment by th0ma7** [comment]
@hgy59 I've been over all your comments, thnx very much for that.
I would suggest we merge this PR prior to yours #4320 ?
What do you think?

**Comment by hgy59** [comment]
> I would suggest we merge this PR prior to yours #4320 ?

Agreed ✔️

**Comment by th0ma7** [comment]
Note that I had already set the groundwork for your PR such as in `spksrc.cross-meson-env.mk` with:
```
ifeq ($(findstring $(ARCH),hi3535),$(ARCH))
  CONFIGURE_ARGS += --cross-file $(MESON_CFG_DIR)/armv7l.cfg
endif
```
Before merging I'll wait for the github-action to complete to confirm all is still ok with the recent changes (besides the known build errors on `ppc853x-5.2` due to late unsupported flags)

**Pull request merged**